### PR TITLE
basic/hashmap: add cleanup of memory pools

### PR DESCRIPTION
**Pull request by keszybz**
It was dropped (by mistake, I assume) in 89439d4fc0d29f04ac68432fd06ab84bc4e36e20.

Fixes the following valgrind warning:

==22564== HEAP SUMMARY:
==22564==     in use at exit: 8,192 bytes in 2 blocks
==22564==   total heap usage: 243 allocs, 241 frees, 151,905 bytes allocated
==22564==
==22564== 4,096 bytes in 1 blocks are still reachable in loss record 1 of 2
==22564==    at 0x4C2FB6B: malloc (vg_replace_malloc.c:299)
==22564==    by 0x4F08A8C: mempool_alloc_tile (mempool.c:62)
==22564==    by 0x4F08B16: mempool_alloc0_tile (mempool.c:81)
==22564==    by 0x4EF8DE0: hashmap_base_new (hashmap.c:748)
==22564==    by 0x4EF8ED9: internal_hashmap_new (hashmap.c:782)
==22564==    by 0x11045D: test_hashmap_copy (test-hashmap-plain.c:87)
==22564==    by 0x115722: test_hashmap_funcs (test-hashmap-plain.c:914)
==22564==    by 0x10FC9D: main (test-hashmap.c:60)
==22564==
==22564== 4,096 bytes in 1 blocks are still reachable in loss record 2 of 2
==22564==    at 0x4C2FB6B: malloc (vg_replace_malloc.c:299)
==22564==    by 0x4F08A8C: mempool_alloc_tile (mempool.c:62)
==22564==    by 0x4F08B16: mempool_alloc0_tile (mempool.c:81)
==22564==    by 0x4EF8DE0: hashmap_base_new (hashmap.c:748)
==22564==    by 0x4EF8EF8: internal_ordered_hashmap_new (hashmap.c:786)
==22564==    by 0x10A2A0: test_ordered_hashmap_copy (test-hashmap-ordered.c:89)
==22564==    by 0x10F70F: test_ordered_hashmap_funcs (test-hashmap-ordered.c:916)
==22564==    by 0x10FCA2: main (test-hashmap.c:61)
==22564==
==22564== LEAK SUMMARY:
==22564==    definitely lost: 0 bytes in 0 blocks
==22564==    indirectly lost: 0 bytes in 0 blocks
==22564==      possibly lost: 0 bytes in 0 blocks
==22564==    still reachable: 8,192 bytes in 2 blocks
==22564==         suppressed: 0 bytes in 0 blocks

v2:
- check if we are the main thread

I submitted this in #7154, but let's make this a separate PR since it's totally unrelated.

### CODE REVIEW

**Comment by poettering** [comment]
hmm, but what happens if a hashmap is allocated from the main thread (hence the pool is used) and then used by a bg thread, and then we terminate in the main thread? we might destroy what the bg thread is still using... or did i miss anything

**Comment by keszybz** [comment]
But, but, ..., seems you're right unfortunately. It seems a blanket solution like this cannot work, I'll need to rethink this.

**Comment by keszybz** [comment]
Updated.

**Comment by keszybz** [comment]
See the updated commit message.

File: src/basic/hashmap.c [hunk 1]
```diff
@@ -278,6 +280,28 @@ static const struct hashmap_type_info hashmap_type_info[_HASHMAP_TYPE_MAX] = {
...
+        /* Be nice to valgrind */
+
+        /* The pool is only allocated by the main thread, but the memory can
+         * be passed to other threads. Let's clean up if we are the main thread
+         * and no other threads are live. */
+        if (!is_main_thread())
+                return;
+
+        r = get_proc_field("/proc/self/status", "Threads", WHITESPACE, &t);
+        if (r < 0 || !streq(t, "1"))
```
Comment (poettering) [inline src/basic/hashmap.c:297]: maybe turn this into a function of its own in process-util.h? i.e. get_n_threads() or so?

**Comment by poettering** [comment]
lgtm

**Comment by keszybz** [comment]
`get_n_threads` would have to return a number, but here we don't really care, so we can avoid parsing. If there's ever a second user we can split it out easily I think.

**Pull request merged**